Feature: serialize UUIDs natively

necaris opened the pull request:
Closes #41 

This is an initial, clumsy, incorrect implementation -- I'm not even sure how to run the tests at this point, and I know that converting the 128-bit `self.int` value in the `UUID` can't be as easy as calling one C function.

But I'd really appreciate some :eyes: and review while I figure out the rest (tips on what to do also appreciated ;-))

ijl commented:
Ok, thanks for contributing. I think the approach is correct.  

`PyLong_AsUnsignedLongLong` returns a 64-bit integer, so rather than construct the equivalent in Rust, I think you need to use `PyObject_Str()`. That or read `PyLongObject` directly to get a u128 in Rust if you find it interesting (it's probably not a meaningful difference?) but that should at least wait to be done as an optimization.

`UUID_PTR` could be looked up in a function for more readability, and if a libpython function accepts a string without a length parameter it needs to be terminated with `\0`.

There are tests in `test_memory.py` using UUIDs that should be modified. You could create a new test module just for UUIDs as well. I would want to test every UUID version and way of creating them. rapidjson tests that the `str` returned by `__str__` is a 32 or 36 chars long (in this case only 36). I don't know--can you break that? It's possible the check is unnecessary but I'm not familiar and would assume it's needed unless you know why not.

This is also a breaking change and I'm not really interested in a v3 for one feature with dataclasses at least also pending, so I suppose add `orjson.SERIALIZE_UUID`? The documentation needs to specify that this is the "canonical" representation and give examples.

It's a lot of detail so it's quite alright if that's too much for your interest.

necaris commented:
Thanks @ijl! I'll start poking around -- I'm hoping I'll learn enough from doing this that I'm able to help more meaningfully in the future :-)

necaris commented:
An update: I have an initial rough implementation at this point that seems to work, and passes the rudimentary tests I've added. 

I also ran a couple of benchmarks to validate this effort -- it looks like natively serializing UUIDs with this approach is about 75% faster on my laptop than calling `str()`: https://gist.github.com/necaris/cb50b25189f155cc4f3fcd1c59e9f28a

I have yet to:
 - [ ] look up `UUID_PTR` in a function for more readability
 - [ ] update `test_memory.py` tests using UUIDs
 - [ ] add tests for other ways of creating UUIDs, and tests for length of the returned string
 - [ ] add documentation for what OPT_SERIALIZE_UUID does, clarify that this is the "canonical" representation and _de_-serialization does _not_ work (unlike `rapidjson`)
 - [ ] add examples of usage

necaris commented:
@ijl I think I've hit all the points you suggested, so think this is ready for review. I've:
 - [x] looked up UUID_PTR in a function for more readability
 - [x] updated tests (`test_memory` and `test_default`) that used UUIDs
 - [x] added tests for other ways of creating UUIDs, and added inline code notes for why the returned string must be 36 characters long
 - [x] added documentation for what `OPT_SERIALIZE_UUID` does, as well as usage examples

Note I went ahead and implemented reading the `PyLongObject` directly in https://github.com/ijl/orjson/pull/48/commits/d8068afb98e40e7b85613d1e8e671ac6a9fb39e5 , which you suggested could be a later optimization. (I'm happy to revert it). It actually resulted in a _substantial_ speedup in my benchmarks. (I updated the gist at https://gist.github.com/necaris/cb50b25189f155cc4f3fcd1c59e9f28a with new results in `results2.txt`).

ijl commented:
Ok. I've looked at why the code calls `PyObject_GetAttr` without checking the error code and why it calls `_PyLong_AsByteArray` and does. I take it there is an assumption of what a UUID object is that I tested in c409737dfadc060110f2f7c89a22a814a25e0172. It looks like `_PyLong_AsByteArray` will never return `-1`. If it did, it would need to be documented that serializing UUIDs could fail and why. To merge the branch I would want to see regression tests for `uuid.uuid1()` etc. and the branch squashed down to one commit.

necaris commented:
@ijl thanks for the review! Your assumptions about `UUID` objects seem correct and I've continued to rely on them -- I've added inline code comments to that effect so anyone looking at `encode.rs` can share those assumptions.

I've added a regression test for all the `uuid*` creator functions, and squashed it down to one commit, also, as requested: https://github.com/ijl/orjson/pull/48/commits/45ac5eafd441cba68d00d753dcaa1dfc93727cb3

ijl commented on src/encode.rs:
```diff
@@ -186,6 +187,41 @@ impl<'p> Serialize for SerializePyObject {
+            };
+            ffi!(Py_DECREF(py_int));
+            let value = u128::from_le_bytes(buffer);
+            let hexadecimal = format!("{:032x}", value);
```
With the `std` feature on SmallVec, this can avoid a heap allocation with:

```
    let mut hexadecimal: SmallVec<[u8; 32]> = SmallVec::with_capacity(32);
    write!(hexadecimal, "{:032x}", value).unwrap();
```

ijl commented on src/encode.rs:
```diff
@@ -186,6 +187,41 @@ impl<'p> Serialize for SerializePyObject {
+                &hexadecimal[20..],
+            ];
+            let string = groups.join("-");
+            serializer.serialize_str(str_from_slice!(string.as_ptr(), string.len()))
```
This could be `serializer.serialize_str(string.as_str())`.

ijl commented:
The tests should be green but for the `win_python38` build having a problem due to an Azure Pipelines change.

necaris commented:
@ijl I'm happy to go back & add to this, address the comments + add the tests. I'll try and get that sorted out in the next couple of hours.

ijl commented:
Ok, if you like.

necaris commented:
@ijl thank you again for the review -- I've updated the PR with the optimizations you suggested and the additional test.

necaris commented:
@ijl thank you again for the review -- I've updated the PR with the optimizations you suggested and the additional test. It's not building correctly at the moment, I'm in the process of fixing it and then I'll have an updated commit.

ijl commented:
@necaris the SmallVec dependency needs `features = ["union", "specialization", "std"]` and the module using it needs the `std::io::Write` trait.

necaris commented:
@ijl I tracked those down, thanks -- also had to add another `SmallVec` for the formatted string since `.join()` or `format!()` on `&[u8]` doesn't work and I wasn't sure what to do otherwise. Based on my rough benchmarking it's another 30% faster, though.

Added tests & rebased down to one commit again.

ijl commented:
Thank you, I modified it a bit on master, including using a separate function, partly to enable #40.

necaris commented:
Thanks @ijl !
